_episodes/04-networks-are-like-onions.md

```diff
@@ -0,0 +1,544 @@
+
+The solution is that we make the network learn in a 'smart' way. The features that we learn should be similar both for small and large images, and similar features (e.g. edges, corners) can appear anywhere in the image (in mathematical terms: *translation invariant*). We do this by making use of a concepts from image processing that precede Deep Learning.
+
+A **convolution matrix**, or **kernel**, is a matrix transformation that we 'slide' over the image to calculate features at each position of the image. For each pixel, we calculate the matrix product between the kernel and the pixel with its surroundings. A kernel is typically small, between 3x3 and 7x7 pixels. We can for example think of the 3x3 kernel:
+~~~
+[[-1, -1, -1],
+ [0, 0, 0]
+ [1, 1, 1]]
+~~~
+{: .language-python}
+This kernel will give a high value to a pixel if it's on a horizontal border between dark and light areas.
+Note that for RGB images, the kernel should also have a depth of 3.
+
+In the following image, we see the effect of such a kernel on the values of a single-channel image. The red cell in the output matrix is the result of multiplying and summing the values of the red square in the input, and the kernel. Applying this kernel to a real image shows that it indeed detects horizontal edges.
+![Convolution on matrix](../fig/04_conv_matrix.png)
+![Convolution on image](../fig/04_conv_image.png)
+
+In our **convolutional layer** our hidden units are a number of convolutional matrices (or kernels), where the values of the matrices are the weights that we learn in the training process. The output of a convolutional layer is an 'image' for each of the kernels, that gives the output of the kernel applied to each pixel.
+
+> ## Border pixels
+>
+> What, do you think, happens to the border pixels when applying a convolution?
+>
+> > ## Solution
+> >
+> > There are different ways of dealing with border pixels.
+> > You can ignore them, which means that your output image is slightly smaller then your input.
+> > It is also possible to 'pad' the borders, e.g. with the same value or with zeros, so that the convolution can also be applied to the border pixels.
+> > In that case, the output image will have the same size as the input image.
+> {: .solution}
+{: .challenge}
+
+> ## Number of model parameters
+>
+> Suppose we apply a convolutional layer with 100 kernels of size 3 * 3 * 3 (the last dimension applies to the rgb channels) to our images of 32 * 32 * 3 pixels. How many parameters do we have? Assume, for simplicity, that the kernels do not use bias terms. Compare this to the answer of the previous exercise
+>
+> > ## Solution
+> >
+> > We have 100 matrices with 3 * 3 * 3 = 27 values each so that gives 27 * 100 = 2700 weights. This is a magnitude of 100 less than the fully connected layer with 100 units!
+> {: .solution}
+{: .challenge}
+
```

Comment:
I would mention/reitterate something about the performance or expressiveness of convolutional layers or allude to it for later. Because now I had the though: "Ok it's a lot smaller, but that doesn't mean much"